### PR TITLE
Separate SessionContext from SessionSupplier

### PR DESCRIPTION
**Pull request by rongrong**

### CODE REVIEW

**Comment by electrum** [comment]
It seems strange to have `SessionContext` be an interface. Why not make it an immutable object with a builder?

File: presto-tests/src/test/java/com/facebook/presto/execution/TestingSessionContext.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    public Map<String, String> getSystemProperties()
+    {
+        return session.getSystemProperties();
+    }
+
+    @Override
+    public Map<String, Map<String, String>> getCatalogSessionProperties()
+    {
+        ImmutableMap.Builder<String, Map<String, String>> catalogSessionProperties = ImmutableMap.builder();
+        session.getConnectorProperties().forEach((key, value) -> catalogSessionProperties.put(key.getCatalogName(), value));
```
Comment (electrum) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestingSessionContext.java]: This isn't doing anything special (such as transform) before calling `forEach`, so write it as a normal for-each loop

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            AccessControl accessControl,
+            SessionPropertyManager sessionPropertyManager)
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.sessionPropertyManager = requireNonNull(sessionPropertyManager, "sessionPropertyManager is null");
+    }
+
+    @Override
+    public Session createSession(
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: These are short, no need to wrap

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            QueryId queryId,
+            SessionContext context)
+    {
+        Identity identity = context.getIdentity();
+        accessControl.checkCanSetUser(identity.getPrincipal().orElse(null), identity.getUser());
+
+        Session.SessionBuilder sessionBuilder = Session.builder(sessionPropertyManager)
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: Static import `SessionBuilder`

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        if (context.getTimeZoneId() != null) {
+            sessionBuilder.setTimeZoneKey(getTimeZoneKey(context.getTimeZoneId()));
+        }
+
+        if (context.getLanguage() != null) {
+            sessionBuilder.setLocale(Locale.forLanguageTag(context.getLanguage()));
+        }
+
+        for (Map.Entry<String, String> entry : context.getSystemProperties().entrySet()) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: Import `Entry`

**Comment by rongrong** [comment]
I don't like SessionContext being an interface either. I made it an interface just because in tests the context factory was created from test sessions. It's weird to have a constructor for SessionContext to take a session in non-test scenarios. So I made it an interface just for that. If we create a SessionContextBuilder, it will only be used in creating test session context as well. But conceptually it does seem better. Do you want me to do that? @electrum

**Comment by electrum** [comment]
I think I am in agreement, but let’s chat on Monday to make sure we’re thinking the same thing.

**Comment by electrum** [comment]
Feel free to merge this now and we can do an additional refactor later.

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java (hunk 3 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: We need to import Session anyways. Is that really necessary?

**Pull request closed without merging**